File: src/python/cgi/rapi.py
#!/usr/bin/env python

# Import library
import os
import sys
import json
import re
import cgi

# Declare function
def parser_request_info():
    data = ""
    if os.environ['CONTENT_LENGTH'] != "":
        data = json.loads(sys.stdin.read())
    return {
        'REQUEST_URI': os.environ['REQUEST_URI'],
        'REQUEST_METHOD': os.environ['REQUEST_METHOD'],
        'SCRIPT_FILENAME': os.environ['SCRIPT_FILENAME'],
        'SCRIPT_NAME': os.environ['SCRIPT_NAME'],
        'QUERY_STRING': os.environ['QUERY_STRING'],
        'CONTENT_TYPE': os.environ['CONTENT_TYPE'],
        'CONTENT_LENGTH': os.environ['CONTENT_LENGTH'],
        'CONTENT': data
    }

def parser_restful_info(url):
    # Regex for get the related fields from URL
    url_pattern = r'/(?P<api>\w+)/(?P<id>[0-9]+)/(?P<name>\w+)$'
    return re.match(url_pattern, url).groupdict()

def response(status, message):
    result = {
        "status": status,
        "message": message
    }
    print "Content-type: application/json"
    print ""
    print json.dumps(result)

def call_get(info):
    return {
        "text": "Call GET method, and retrieve information.",
        "body": info
    }

def call_post(info):
    return {
        "text": "Call POST method, new information with body.",
        "body": info["req"]["CONTENT"]
    }

def call_put(info):
    return {
        "text": "Call PUT method, update information with body.",
        "body": info["req"]["CONTENT"]
    }

def call_del(info):
    return "Call DELETE method, and remove information."

switch = {"GET": call_get, "POST": call_post, "PUT": call_put, "DELETE": call_del}
# Execute script
try:
    req = parser_request_info()
    info = {
        "restful": parser_restful_info(req["SCRIPT_NAME"]),
        "req": req
    }
    response("success", switch[req["REQUEST_METHOD"]](info))
except Exception as e:
    response("error", e.args[0])

File: src/perl/cgi/rapi.pl
#!/usr/bin/env perl

# Import library
use JSON;
use Switch;

# Declare function
sub parser_request_info {
    $CONTNET = "";
    $CONTENT_TYPE = $ENV{'CONTENT_TYPE'};
    $CONTENT_LENGTH = $ENV{'CONTENT_LENGTH'};
    if ( ! $CONTENT_LENGTH eq "" ) {
        read (STDIN, $CONTENT, $ENV{'CONTENT_LENGTH'});
    }
    %req=(
        "REQUEST_METHOD"=>$ENV{'REQUEST_METHOD'},
        "REQUEST_URI"=>$ENV{'REQUEST_URI'},
        "SCRIPT_FILENAME"=>$ENV{'SCRIPT_FILENAME'},
        "SCRIPT_NAME"=>$ENV{'SCRIPT_NAME'},
        "QUERY_STRING"=>$ENV{'QUERY_STRING'},
        "CONTENT_TYPE"=>$CONTENT_TYPE,
        "CONTENT_LENGTH"=>$CONTENT_LENGTH,
        "CONTENT"=>$CONTENT
    );
    return \%req
}

sub parser_restful_info {
    my($url) = @_;
    # Regex for get the related fields from URL
    $url =~ m/(\w+)\/(\w+)\/(\w+)/;
    %re = (
        "api"=>$1,
        "id"=>$2,
        "name"=>$3
    );
    return \%re;
}

sub response {
    my($status, $message) = @_;
    %obj=(
        "status"=>$status,
        "message"=>$message
    );
    $json_text_obj = encode_json \%obj;
    print "Content-type: application/json";
    print "\n\n";
    print "$json_text_obj\n";
}

sub call_get {
    my($info) = @_;
    %ret = (
        "text"=>"Call GET method, and retrieve information.",
        "body"=>$info
    );
    return \%ret;
}

sub call_post {
    my($info) = @_;
    $json_obj_ref = decode_json $info{"req"}{"CONTENT"};
    %ret = (
        "text"=>"Call POST method, new information with body.",
        "body"=>$json_obj_ref
    );
    return \%ret;
}

sub call_put {
    my($info) = @_;
    $json_obj_ref = decode_json $info{"req"}{"CONTENT"};
    %ret = (
        "text"=>"Call PUT method, update information with body.",
        "body"=>$json_obj_ref
    );
    return \%ret;
}

sub call_del {
    my($info) = @_;
    return "Call DELETE method, and remove information."
}
# Execute script
$req = &parser_request_info();
%info = (
    "restful"=>&parser_restful_info($req{"SCRIPT_NAME"}),
    "req"=>$req
);
switch($req{"REQUEST_METHOD"}) {
    case "GET"    { $result = &call_get(\%info); }
    case "POST"    { $result = &call_post(\%info); }
    case "PUT"    { $result = &call_put(\%info); }
    case "DELETE"    { $result = &call_del(\%info); }
}
&response("success", $result);
